src/charm.py

```diff
@@ -4,101 +4,173 @@
+"""MySQL-Router charm."""
 
+import json
 import logging
+from typing import Any, Dict
 
 from ops.charm import CharmBase
-from ops.framework import StoredState
 from ops.main import main
-from ops.model import ActiveStatus
+from ops.model import ActiveStatus, WaitingStatus
 
 logger = logging.getLogger(__name__)
+DATABASE = "database"
+PEER = "mysql-router"
 
 
-class OperatorTemplateCharm(CharmBase):
+class MysqlRouterOperatorCharm(CharmBase):
     """Charm the service."""
 
-    _stored = StoredState()
-
     def __init__(self, *args):
         super().__init__(*args)
-        self.framework.observe(self.on.httpbin_pebble_ready, self._on_httpbin_pebble_ready)
+
+        self.name = "mysqlrouter"
+
+        self.framework.observe(self.on.mysqlrouter_pebble_ready, self._on_mysqlrouter_pebble_ready)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
-        self.framework.observe(self.on.fortune_action, self._on_fortune_action)
-        self._stored.set_default(things=[])
+        self.framework.observe(
+            self.on.database_relation_created, self._on_database_relation_created
+        )
+        self.framework.observe(
+            self.on.database_relation_departed, self._on_database_relation_departed
+        )
+
+    @property
+    def has_db(self) -> bool:
+        """Only consider a DB connection if we have config info."""
+        rel = self.model.get_relation(DATABASE)
+        return len(rel.units) > 0 if rel is not None else False
+
+    @property
+    def peers(self) -> list:
+        """Fetch the peer relation."""
+        return self.model.get_relation(PEER)
+
+    def set_peer_data(self, key: str, data: Any) -> None:
+        """Put information into the peer data bucket instead of `StoredState`."""
+        if self.unit.is_leader():
+            self.peers.data[self.app][key] = json.dumps(data)
+
+    def get_peer_data(self, key: str) -> Any:
+        """Retrieve information from the peer data bucket instead of `StoredState`."""
+        data = self.peers.data[self.app].get(key, "")
+        return json.loads(data) if data else {}
+
+    def _configure(self) -> None:
+        """Configure the charm."""
+        # Get the relation data
+        data = self.get_peer_data(DATABASE)
+
+        config_data = json.loads(data["mysql"])
+
+        if not self._validate_config(config_data):
+            logger.error("Invalid config")
+            self.unit.status = WaitingStatus("Invalid relation config")
+            return
+
+        # Define Pebble layer configuration
+        pebble_layer = self._mysqlrouter_layer(
+            port=config_data["port"],
+            host=config_data["host"],
+            user=config_data["user"],
+            password=config_data["password"],
+        )
+        # Add initial Pebble config layer using the Pebble API
+        container = self.unit.get_container(self.name)
+        plan = container.get_plan()
+
+        if plan.services != pebble_layer["services"]:
```

Comment:
If `_mysqlrouter_layer` returns a `Layer` class, you can modify this to:
```python
if plan.services != pebble_layer.services:
```

Reply:
Work as a charm (pun intended)